### ISSUE TITLE
automatically try relay 

### ISSUE THREAD

**Issue by whyrusleeping** [ipfs/go-ipfs#4213]
Now that we have relay support its time to start really solving problems with it. It seems to me that the first big thing to look at here is automatically trying to connect via relay if normal methods fail. This is hard for a variety of reasons. We dont want to use a relay if we can possibly connect directly, but we dont want to sit around waiting forever for dial attempts to fail. We also then have to detect if using relay is even an option. For that to be the case we have to know that our target supports relay STOP, and that we are connected to a relay HOP peer that is connected to them.

Once we have a relayed connection to a peer, we should ask them to attempt making a direct connection to us, and if they are successful, use that instead of the relayed connection. This implies we need a good method of migrating our 'chosen' connection to another peer over to another Conn.

There is a lot here, but I wanted to open up the discussion so we can make a plan of attack on getting this all working.

cc @vyzo @Stebalien @diasdavid @Kubuxu @dryajov

**Comment by dryajov** [comment]
Glad to see this issue! 👍 

- Currently we have the `CAN_HOP` message which allows discovering HOPs but not STOPs, being able to do that would be supper useful. I've seen several proposals for both relay capabilities discovery as well as a more general capabilities discovery - it be nice to consolidate those, I'll search around and try to link those issues here.

> We dont want to use a relay if we can possibly connect directly, but we dont want to sit around waiting forever for dial attempts to fail

- This is how I currently handle it in JS:
   - When a `/p2p-circuit` address is being dialed
       - if there are other transports for that peer that we know of, then those are tried first, falling back to circuit if all attempts to dial have failed
       - if there are no dialable transports at all for that peer, then `circuit` is used
    - When a non `/p2p-circuit` address is dialed, circuit is used only as a fallback when all other transports have failed.

Just wanted to document this here, as a possible approach. There also might be some differences on how the GO and the JS implementations currently handle this cases - it is something to investigate and possibly sync up across both implementation, but I know that the core cases should work for both, which should be good enough for now? I believe this can use a separate issue as well.

- On connecting directly 
   - Do we currently have any way to upgrade connections? Would this even qualify as a  connection upgrade? 
       - One approach could be, to dial the peer again after a circuit connection has been established, after asking identify (or something else?) for addition info on the peer, if that connection succeeds, use it, otherwise keep using the relayed connection. Just some thoughts.

**Comment by dryajov** [comment]
Opened this issue to talk about implementation differences - https://github.com/libp2p/js-libp2p-circuit/issues/16

**Comment by dryajov** [comment]
Some prior discussions and relevant issues:
- Relay discovery: https://github.com/libp2p/js-libp2p-circuit/issues/11
- P2P Circuit Advertisements: https://github.com/libp2p/specs/issues/33
- Service Discovery instead of Protocol Negotiation: https://github.com/libp2p/specs/issues/32
- Proposal: Add a `CAN_HOP` message Type to circuit-relay to allow passive relay discovery: https://github.com/libp2p/specs/issues/27

**Comment by vyzo** [comment]
> For that to be the case we have to know that our target supports relay STOP, and that we are connected to a relay HOP peer that is connected to them.

I don't think we can query this with a `CAN_HOP` style message, as we need to have a working connection to the peer or a HOP peer that is connected to them in order to test it.
So it seems to me that we need to solve the relay address advertisement issue before we can address this.

**Comment by lidel** [comment]
Closing as we've removed Relays V1 from Kubo and replaced them with [Circuit Relay v2](https://github.com/libp2p/specs/blob/master/relay/circuit-v2.md). Kubo will use them automatically, if needed.
Any issues related to relays should be filled as new tickets,  against Kubo >0.19.2.
